File: backend/workline/knowledge/cache/persistent.py
"""Thread-safe L2 Persistent filesystem cache under .workline/cache/"""

import json
import os
import shutil
import threading
import time
from typing import Any, Dict, Optional, Tuple
from backend.workline.knowledge.cache.models import CacheMetadata, CacheObjectType


class PersistentCache:
    """Stores serialized JSON cache items in organized directories under .workline/cache/"""

    def __init__(self, base_dir: str = ".workline/cache"):
        self.base_dir = base_dir
        self._lock = threading.RLock()
        self._ensure_directories()

    def _ensure_directories(self) -> None:
        try:
            subdirs = ["metadata", "documents", "embeddings", "retrieval", "context", "summaries"]
            for sub in subdirs:
                os.makedirs(os.path.join(self.base_dir, sub), exist_ok=True)
        except Exception:
            pass

    def _get_path(self, key: str, object_type: CacheObjectType) -> str:
        safe_key = key.replace(":", "_").replace("/", "_").replace("\\", "_")
        folder = "metadata"
        if object_type in (CacheObjectType.DOCUMENT_PARSE, CacheObjectType.DOCUMENT_CHUNK):
            folder = "documents"
        elif object_type == CacheObjectType.EMBEDDING:
            folder = "embeddings"
        elif object_type == CacheObjectType.RETRIEVAL:
            folder = "retrieval"
        elif object_type == CacheObjectType.CONTEXT:
            folder = "context"
        elif object_type in (CacheObjectType.SUMMARY, CacheObjectType.RESEARCH):
            folder = "summaries"

        return os.path.join(self.base_dir, folder, f"{safe_key}.json")

    def get(self, key: str, object_type: CacheObjectType) -> Optional[Tuple[CacheMetadata, Any]]:
        with self._lock:
            path = self._get_path(key, object_type)
            if not os.path.exists(path):
                return None

            try:
                with open(path, "r", encoding="utf-8") as f:
                    payload = json.load(f)

                meta = CacheMetadata.model_validate(payload["metadata"])
                data = payload["data"]

                # Check expiration
                if meta.expires_at > 0 and time.time() > meta.expires_at:
                    self.delete(key, object_type)
                    return None

                return meta, data
            except Exception:
                # Discard corrupted file
                self.delete(key, object_type)
                return None

    def set(self, key: str, meta: CacheMetadata, data: Any) -> None:
        with self._lock:
            path = self._get_path(key, meta.object_type)
            try:
                payload = {
                    "metadata": meta.model_dump(),
                    "data": data,
                }
                with open(path, "w", encoding="utf-8") as f:
                    json.dump(payload, f, ensure_ascii=False)
            except Exception:
                pass

    def delete(self, key: str, object_type: CacheObjectType) -> bool:
        with self._lock:
            path = self._get_path(key, object_type)
            if os.path.exists(path):
                try:
                    os.remove(path)
                    return True
                except Exception:
                    return False
            return False

    def clear(self) -> None:
        with self._lock:
            if os.path.exists(self.base_dir):
                try:
                    shutil.rmtree(self.base_dir)
                except Exception:
                    pass
            self._ensure_directories()

    def get_stats(self) -> Tuple[int, int]:
        """Returns (total_entries, total_size_bytes)."""
        with self._lock:
            total_entries = 0
            total_size = 0
            for root, _, files in os.walk(self.base_dir):
                for f in files:
                    if f.endswith(".json"):
                        total_entries += 1
                        try:
                            total_size += os.path.getsize(os.path.join(root, f))
                        except Exception:
                            pass
            return total_entries, total_size

File: packages/knowledge/cache/persistent.ts
/**
 * L2 Persistent File-System Cache under .workline/cache/
 */

import * as fs from "fs";
import * as path from "path";
import { CacheEntry, CacheObjectType } from "./metadata";
import { CacheSerializer } from "./serializer";

export class PersistentCache {
  private readonly baseDir: string;
  private readonly maxSizeBytes: number;

  constructor(baseDir: string = ".workline/cache", maxSizeBytes: number = 512 * 1024 * 1024) {
    this.baseDir = baseDir;
    this.maxSizeBytes = maxSizeBytes;
    this.ensureDirectoryStructure();
  }

  private ensureDirectoryStructure(): void {
    const subdirs = ["metadata", "documents", "embeddings", "retrieval", "context", "summaries"];
    for (const sub of subdirs) {
      const p = path.join(this.baseDir, sub);
      if (!fs.existsSync(p)) {
        fs.mkdirSync(p, { recursive: true });
      }
    }
  }

  private getFilePath(key: string, objectType: CacheObjectType): string {
    const safeKey = key.replace(/[^a-zA-Z0-9_-]/g, "_");
    let folder = "metadata";
    switch (objectType) {
      case CacheObjectType.DOCUMENT_PARSE:
      case CacheObjectType.DOCUMENT_CHUNK:
        folder = "documents";
        break;
      case CacheObjectType.EMBEDDING:
        folder = "embeddings";
        break;
      case CacheObjectType.RETRIEVAL:
        folder = "retrieval";
        break;
      case CacheObjectType.CONTEXT:
        folder = "context";
        break;
      case CacheObjectType.SUMMARY:
        folder = "summaries";
        break;
    }
    return path.join(this.baseDir, folder, `${safeKey}.json`);
  }

  public get<T>(key: string, objectType: CacheObjectType): CacheEntry<T> | null {
    try {
      const filePath = this.getFilePath(key, objectType);
      if (!fs.existsSync(filePath)) return null;

      const content = fs.readFileSync(filePath, "utf-8");
      const entry = CacheSerializer.deserialize<T>(content);
      if (!entry) {
        // Discard corrupted entry
        this.delete(key, objectType);
        return null;
      }

      // Check TTL expiration
      if (entry.metadata.expiresAt > 0 && Date.now() > entry.metadata.expiresAt) {
        this.delete(key, objectType);
        return null;
      }

      return entry;
    } catch {
      return null;
    }
  }

  public set<T>(entry: CacheEntry<T>): void {
    try {
      const filePath = this.getFilePath(entry.metadata.cacheKey, entry.metadata.objectType);
      const serialized = CacheSerializer.serialize(entry);
      fs.writeFileSync(filePath, serialized, "utf-8");
    } catch {
      // Graceful fallback on disk write failure
    }
  }

  public delete(key: string, objectType: CacheObjectType): boolean {
    try {
      const filePath = this.getFilePath(key, objectType);
      if (fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
        return true;
      }
      return false;
    } catch {
      return false;
    }
  }

  public clear(): void {
    try {
      if (fs.existsSync(this.baseDir)) {
        fs.rmSync(this.baseDir, { recursive: true, force: true });
        this.ensureDirectoryStructure();
      }
    } catch {}
  }

  public getStats(): { totalEntries: number; totalSizeBytes: number } {
    let totalEntries = 0;
    let totalSizeBytes = 0;

    const countDir = (dir: string) => {
      if (!fs.existsSync(dir)) return;
      const items = fs.readdirSync(dir);
      for (const item of items) {
        const fullPath = path.join(dir, item);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
          countDir(fullPath);
        } else if (stat.isFile() && item.endsWith(".json")) {
          totalEntries++;
          totalSizeBytes += stat.size;
        }
      }
    };

    countDir(this.baseDir);
    return { totalEntries, totalSizeBytes };
  }
}
